subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -238,18 +240,71 @@ static uint8_t l2cap_br_get_ident(void)
 	return ident;
 }
 
+static void raise_data_ready(struct bt_l2cap_br_chan *br_chan)
+{
+	if (!atomic_set(&br_chan->_pdu_ready_lock, 1)) {
+		sys_slist_append(&br_chan->chan.conn->upper_data_ready,
+				 &br_chan->_pdu_ready);
+		LOG_DBG("data ready raised");
+	} else {
+		LOG_DBG("data ready already");
+	}
+
+	bt_conn_data_ready(br_chan->chan.conn);
+}
+
+static void lower_data_ready(struct bt_l2cap_br_chan *br_chan)
+{
+	struct bt_conn *conn = br_chan->chan.conn;
+	sys_snode_t *s = sys_slist_get(&conn->upper_data_ready);
+
+	__ASSERT_NO_MSG(s == &br_chan->_pdu_ready);
+	(void)s;
+
+	atomic_t old = atomic_set(&br_chan->_pdu_ready_lock, 0);
+
+	__ASSERT_NO_MSG(old);
+	(void)old;
+}
+
+static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+{
+	struct bt_conn *conn = br_chan->chan.conn;
+
+	sys_slist_find_and_remove(&conn->upper_data_ready,
+				  &br_chan->_pdu_ready);
+
+	atomic_set(&br_chan->_pdu_ready_lock, 0);
+}
+
 int bt_l2cap_br_send_cb(struct bt_conn *conn, uint16_t cid, struct net_buf *buf,
 			bt_conn_tx_cb_t cb, void *user_data)
 {
 	struct bt_l2cap_hdr *hdr;
+	struct bt_l2cap_chan *ch = bt_l2cap_br_lookup_tx_cid(conn, cid);
+	struct bt_l2cap_br_chan *br_chan = CONTAINER_OF(ch, struct bt_l2cap_br_chan, chan);
 
-	LOG_DBG("conn %p cid %u len %zu", conn, cid, buf->len);
+	LOG_DBG("chan %p buf %p len %zu", br_chan, buf, buf->len);
 
 	hdr = net_buf_push(buf, sizeof(*hdr));
 	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
-	hdr->cid = sys_cpu_to_le16(cid);
+	hdr->cid = sys_cpu_to_le16(br_chan->tx.cid);
```

Comment:
We need it for channel lookup, right above. 

I would prefer not touching/refactoring `br` any more than I have to. Part because my employer doesn't have interest in it, but mostly because I can't really test it properly. Or at least in a non time-consuming and manual way.